### PR TITLE
`fn dav1d_decode_frame_init_cdf`: Cleanup and make mostly safe

### PR DESCRIPTION
**Pull request by kkysen**
This involved some heavy iterator code to avoid bounds checks as much as possible.  One especially, 08f1a37caaa6da368108f931ab6b974198528f60, I want to benchmark.  @fbossen, how have you been doing that?

### CODE REVIEW

**Comment by fbossen** [comment]
I don't have `perf stat` available on my ARM system for benchmarking. I simply time the decoding of the four "standard" bitstreams multiple times (3 or more) and pick the best value. I will run a test to compare this PR to the current tip of `main` and see whether there is any significant difference.

**Comment by fbossen** [comment]
Looks like there might be a slight performance regression between this PR and `main`. However, the number of commits (and PRs) between the two is significant. Let me see whether I can pinpoint this regression.

**Review by randomPoison — APPROVED** [review]

File: src/decode.rs [hunk 1]
```diff
@@ -5648,90 +5647,100 @@ pub unsafe extern "C" fn dav1d_decode_frame_init(f: *mut Dav1dFrameContext) -> l
...
+
+    if (*f.frame_hdr).refresh_context != 0 {
+        dav1d_cdf_thread_copy(f.out_cdf.data.cdf, &mut f.in_cdf);
     }
+
+    let uses_2pass = c.n_fc > 1;
+
+    let tiling = &(*f.frame_hdr).tiling;
+
+    let n_file_data = f.n_tile_data.try_into().unwrap();
```
[the remark this answers is not in the document]
Comment (kkysen) [reply src/decode.rs:5662]: Yes, good catch.

**Comment by kkysen** [comment]
> I don't have `perf stat` available on my ARM system for benchmarking. I simply time the decoding of the four "standard" bitstreams multiple times (3 or more) and pick the best value.

Are you running them with filesystem caches full then?

Also, @thedataking, how did you benchmark things with `perf`?

**Comment by kkysen** [comment]
> Looks like there might be a slight performance regression between this PR and `main`. However, the number of commits (and PRs) between the two is significant. Let me see whether I can pinpoint this regression.

I tried benchmarking this PR vs. the current `main` (02e531e) and the previous `main` when you ran them (becc4f0) on `donna` (x86_64) with asm enabled, but I didn't find any differences.  I used `hyperfine` for the benchmarking, which ran 10 runs, and there were only small (±0.1-0.5%) variations in average and min times that weren't consistently reproducible.  Could you benchmark them on your arm machine if you're still seeing the regression?

File: src/decode.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkysen) [reply src/decode.rs]: Fixed in [ba78d5d..a00bd66](https://github.com/memorysafety/rav1d/compare/ba78d5d8337aa71ad7ec42387c95db8fc0a9eeff..a00bd664367f187286b2887b88e9a0e1290744ad).

**Pull request merged**